pixeltable/tests/test_exprs.py

```diff
@@ -356,6 +359,112 @@ def test_arrays(self, test_tbl: catalog.Table) -> None:
+        # Type conversion error
+        status = t.add_column(c2_as_string=t.c2.astype(StringType()))
+        assert status.num_excs == t.count()
+
+    def test_apply(self, test_tbl: catalog.Table) -> None:
+
+        t = test_tbl
+
+        # For each column c1, ..., c5, we create a new column ci_as_str that converts it to
+        # a string, then check that each row is correctly converted
+        # (For c1 this is the no-op string-to-string conversion)
+        for col_id in range(1, 6):
+            col_name = f'c{col_id}'
+            str_col_name = f'c{col_id}_str'
+            status = t.add_column(**{str_col_name: t[col_name].apply(str)})
+            assert status.num_excs == 0
+            data = t.select(t[col_name], t[str_col_name]).collect()
+            for row in data:
+                assert row[str_col_name] == str(row[col_name])
+
+        # For columns c6, c7, try using json.dumps and json.loads to emit and parse JSON <-> str
+        for col_id in range(6, 8):
+            col_name = f'c{col_id}'
+            str_col_name = f'c{col_id}_str'
+            back_to_json_col_name = f'c{col_id}_back_to_json'
+            status = t.add_column(**{str_col_name: t[col_name].apply(json.dumps)})
+            assert status.num_excs == 0
+            status = t.add_column(**{back_to_json_col_name: t[str_col_name].apply(json.loads)})
+            assert status.num_excs == 0
+            data = t.select(t[col_name], t[str_col_name], t[back_to_json_col_name]).collect()
+            for row in data:
+                assert row[str_col_name] == json.dumps(row[col_name])
+                assert row[back_to_json_col_name] == row[col_name]
+
+        def f1(x):
+            return str(x)
+
+        # Now test that a function without a return type throws an exception ...
+        with pytest.raises(RuntimeError):
+            t['c2'].apply(f1)
+
+        # ... but works if the type is specified explicitly.
+        status = t.add_column(c2_str_f1=t['c2'].apply(f1, col_type=StringType()))
+        assert status.num_excs == 0
+
+        # Test that the return type of a function can be successfully inferred.
+        def f2(x) -> str:
+            return str(x)
+
+        status = t.add_column(c2_str_f2=t['c2'].apply(f2))
+        assert status.num_excs == 0
+
+        # Test various validation failures.
+
+        def f3(x, y) -> str:
+            return f'{x}{y}'
+
+        with pytest.raises(RuntimeError):
+            t['c2'].apply(f3)       # Too many required parameters
+
+        def f4() -> str:
+            return "pixeltable"
+
+        with pytest.raises(RuntimeError):
+            t['c2'].apply(f4)       # No positional parameters
+
+        def f5(**kwargs) -> str:
+            return ""
+
+        with pytest.raises(RuntimeError):
+            t['c2'].apply(f5)       # No positional parameters
+
+        # Ensure these varargs signatures are acceptable
+
+        def f6(x, **kwargs) -> str:
```

Comment:
I think support kwargs (eg, apply(json.loads, parse_float=...) would be useful. At least leave a todo in apply()

Reply:
Aha. So `expr.apply(json.loads, parse_float=...)` would call `json.loads(x, parse_float=...)` on each element x?

Note the "functional" way of doing this would be

`expr.apply(lambda x: json.loads(x, parse_float=...))`

or

```
def my_loader(x): json.loads(x, parse_float=...)
expr.apply(my_loader)
```

either of which of course already works. But your suggestion feels more "pythonic" and propagates the function name in a more concise way.

On the other hand, the lambda pattern is more general, since it also handles situations involving additional positional parameters, e.g.

```
def f(a, b): return a + b
expr.apply(lambda x: f(x, 3))
```

Of course we can also envision ways of handling this without the lambda by further extending apply().

I'll leave a TODO for now and we can ruminate on it some more.